concordium-consensus/src/Concordium/ImportExport.hs

```diff
@@ -116,30 +125,122 @@ sectionHeaderLength :: Word64
         else do
             hdl <- openBinaryFile p WriteMode
             return (p, hdl)
   where
+    -- The old, potentially unversionsioned, file extension
+    oldExt :: String
     oldExt = takeExtensions p
-    newExt =
+    -- The next versioned file extension
+    getNewExt :: IO String
+    getNewExt =
         if oldExt == takeExtension p
-            then ".2" ++ oldExt
-            else
-                (\x -> "." ++ x ++ ".dat")
-                    . show
-                    . (+ (1 :: Int))
-                    . read
-                    . (!! 1)
-                    . splitOn "."
-                    $ oldExt
+            then return $ ".2" ++ oldExt
+            else case tryGetVersionNumber (T.pack oldExt) of
+                Left err -> fail err
+                Right v -> return $ "." <> show (v + 1) <> ".dat"
+    -- Given a file name attempt to get its version number
+    tryGetVersionNumber :: T.Text -> Either String Integer
+    tryGetVersionNumber = AP.parseOnly $ do
+        AP.skipMany $ AP.skip (AP.inClass "a-zA-Z0-9-")
+        _ <- AP.char '.'
+        AP.decimal
+
+-- |Data type used to represent a line with genesis block hash in the block index file.
+type BlockIndexGenesisHash = BlockHash
+
+-- |Data type used to represent a line with chunk information in the block index file.
+data BlockIndexChunkInfo = BlockIndexChunkInfo
+    { filename :: T.Text,
+      genesisIndex :: GenesisIndex,
+      blockHeightFirst :: BlockHeight,
+      blockHeightLast :: BlockHeight
+    }
+    deriving (Show)
+
+-- |Data type used to represent the block index file.
+type BlockIndex = Seq (BlockIndexGenesisHash, Seq BlockIndexChunkInfo)
+
+-- |Parse a line of the block index file containing block genesis data.
+parseGenesisDataLine :: AP.Parser BlockIndexGenesisHash
+parseGenesisDataLine = do
+    _ <- AP.string (T.pack "# genesis hash ")
+    hash <- AP.count 64 $ AP.satisfy isHexDigit
+    AP.skip AP.isEndOfLine
+    return $ read hash
+
+-- |Parse a line of the block index file containing chunk information.
+parseChunkLine :: AP.Parser BlockIndexChunkInfo
+parseChunkLine = do
+    filename <- AP.takeWhile1 (AP.inClass ".a-zA-Z0-9-")
+    _ <- AP.char ','
+    genesisIndex <- AP.decimal
+    _ <- AP.char ','
+    blockHeightStart <- AP.decimal
+    _ <- AP.char ','
+    blockHeightEnd <- AP.decimal
+    AP.skip AP.isEndOfLine
+    return $ BlockIndexChunkInfo filename genesisIndex blockHeightStart blockHeightEnd
+
+-- | Parse a block index file.
+parseBlockIndexFile :: AP.Parser BlockIndex
+parseBlockIndexFile = do
+    atEnd <- AP.atEnd
+    if not atEnd
+        then do
+            list <-
+                AP.manyTill
+                    ( do
+                        genesisData <- parseGenesisDataLine
+                        chunks <- AP.many' parseChunkLine
+                        return (genesisData, fromList chunks)
+                    )
+                    AP.endOfInput
```

Comment:
Which case does just
```haskell

                AP.manyTill
                    ( do
                        genesisData <- parseGenesisDataLine
                        chunks <- AP.many' parseChunkLine
                        return (genesisData, fromList chunks)
                    )
                    AP.endOfInput
```
not handle? Why do you have to check whether we are at the end of the stream manually ?